exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,490 @@
+    pub fn new(code: u8) -> Self {
+        Self {
+            code,
+            description: None,
+        }
+    }
+
+    /// Constructs a new `ExecutionError` instance with the given error code and description.
+    pub fn with_description(code: u8, description: String) -> Self {
+        Self {
+            code,
+            description: Some(description),
+        }
+    }
+}
+
+/// Type of the transaction error.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum TransactionErrorType {
+    /// Panic occurred during transaction execution.
+    Panic,
+    /// User-defined error-code. Can have different meanings for different transactions and
+    /// services.
+    Code(u8),
+}
+
+/// Extended by the framework result of unsuccessful transaction execution.
+///
+/// # Notes:
+///
+/// - Content of `description`' field is excluded from hash calculation (see `StorageValue`
+///   implementation for the details).
+/// - `TransactionErrorType::Panic` is set by the framework if panic is raised during transaction
+///   execution.
+/// - `TransactionError` implements `Display` which can be used for obtaining a simple error
+///   description.
+///
+/// # Examples
+///
+/// ```
+/// # use exonum::blockchain::{ExecutionError, TransactionError};
+///
+/// # let transaction_error: TransactionError = ExecutionError::new(1).into();
+/// // Prints user friendly error description.
+/// println!("Transaction error: {}", transaction_error);
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TransactionError {
+    /// Error type, see `TransactionErrorType` for the details.
+    error_type: TransactionErrorType,
+    /// Optional error description.
+    description: Option<String>,
+}
+
+impl TransactionError {
+    /// Creates a new `TransactionError` instance with the specified error type and description.
+    fn new(error_type: TransactionErrorType, description: Option<String>) -> Self {
+        Self {
+            error_type,
+            description,
+        }
+    }
+
+    /// Creates a new `TransactionError` instance with the specified error code and description.
+    fn code(code: u8, description: Option<String>) -> Self {
+        Self::new(TransactionErrorType::Code(code), description)
+    }
+
+    /// Creates a new `TransactionError` representing panic with the given description.
+    fn panic(description: Option<String>) -> Self {
+        Self::new(TransactionErrorType::Panic, description)
+    }
+
+    /// Creates a new `TransactionError` instance from `std::thread::Result`'s `Err`.
+    pub(crate) fn from_panic(_panic: &Box<Any + Send>) -> Self {
+        // TODO: Try to get description from panic.
+        Self::panic(None)
+    }
+
+    /// Returns error type of this `TransactionError` instance.
+    pub fn error_type(&self) -> TransactionErrorType {
+        self.error_type
+    }
+
+    /// Returns an optional error description.
+    pub fn description(&self) -> &Option<String> {
```

Comment:
Nitpick: maybe, it makes sense to convert this to `Option<&str>` by using `Option::as_ref()` method?